test/test_selectors.py

```diff
@@ -0,0 +1,664 @@
+    def __init__(self, testcase, lower=None, upper=None):
+        self.testcase = testcase
+        self.lower = lower
+        self.upper = upper
+        self.start_time = None
+        self.end_time = None
+
+    def __enter__(self):
+        self.start_time = get_time()
+
+    def __exit__(self, *args, **kwargs):
+        self.end_time = get_time()
+        total_time = self.end_time - self.start_time
+        if self.lower is not None:
+            self.testcase.assertGreaterEqual(total_time, self.lower * (1.0 - LOWER_TOLERANCE))
+        if self.upper is not None:
+            self.testcase.assertLessEqual(total_time, self.upper * (1.0 + UPPER_TOLERANCE))
+
+
+class TimerMixin(object):
+    def assertTakesTime(self, lower=None, upper=None):
+        return TimerContext(self, lower=lower, upper=upper)
+
+
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector")
+class WaitForIOTest(unittest.TestCase, AlarmMixin, TimerMixin):
+    """ Tests for the higher level wait_for_* functions. """
+
+    def make_socketpair(self):
+        rd, wr = socketpair()
+
+        rd.settimeout(0.0)
+        wr.settimeout(0.0)
+
+        self.addCleanup(rd.close)
+        self.addCleanup(wr.close)
+        return rd, wr
+
+    def test_selector_error(self):
+        err = selectors.SelectorError(1)
+        self.assertEqual(err.__repr__(), "<SelectorError errno=1>")
+        self.assertEqual(err.__str__(), "<SelectorError errno=1>")
+
+    def test_wait_for_read_single_socket(self):
+        rd, wr = self.make_socketpair()
+        self.assertEqual([], selectors.wait_for_read(rd, timeout=0.001))
+
+    def test_wait_for_read_multiple_socket(self):
+        rd, rd2 = self.make_socketpair()
+        self.assertEqual([], selectors.wait_for_read([rd, rd2], timeout=0.001))
+
+    def test_wait_for_read_empty(self):
+        self.assertEqual([], selectors.wait_for_read([], timeout=0.001))
+
+    def test_wait_for_write_single_socket(self):
+        wr, wr2 = self.make_socketpair()
+        self.assertEqual([wr], selectors.wait_for_write(wr, timeout=0.001))
+
+    def test_wait_for_write_multiple_socket(self):
+        wr, wr2 = self.make_socketpair()
+        result = selectors.wait_for_write([wr, wr2], timeout=0.001)
+        # assertItemsEqual renamed in Python 3.x
+        if hasattr(self, "assertItemsEqual"):
+            self.assertItemsEqual([wr, wr2], result)
+        else:
+            self.assertCountEqual([wr, wr2], result)
+
+    def test_wait_for_write_empty(self):
+        self.assertEqual([], selectors.wait_for_write([], timeout=0.001))
+
+    def test_wait_for_non_list_iterable(self):
+        rd, wr = self.make_socketpair()
+        iterable = {'rd': rd}.values()
+        self.assertEqual([], selectors.wait_for_read(iterable, timeout=0.001))
+
+    def test_wait_timeout(self):
+        rd, wr = self.make_socketpair()
+        with self.assertTakesTime(lower=0.0, upper=0.1):
+            selectors.wait_for_read([rd], timeout=0.01)
```

Comment:
You're right! I changed all the numerical constants in this test case to be named.